### ISSUE TITLE
Choose checkpoint model in testing step?

### ISSUE THREAD

**Issue by tm1412** [nianlonggu/MemSum#12]
How do you choose a model checkpoint to use for testing or simply the final checkpoint of that model?

**Comment by nianlonggu** [comment]
I used the performance on the validation set (by empirically choose a set of hyparameters, e.g., stopping threshold) to select the model checkpoint that is used for testing

**Comment by tm1412** [comment]
Oh thank you! Can you help me identify where that code snippet is located and the command to execute it?

**Comment by nianlonggu** [comment]
Please check the jupyter notebook:
https://github.com/nianlonggu/MemSum/blob/main/Data_processing_training_and_testing_for_MemSum.ipynb

In the training part, I printed the validation ROUGE performance and use the checkpoint with the highest validation score for testing. This selection is done manually.

**Comment by tm1412** [comment]
As i can see, you have a file named get_optimal_batch.py. Does it have any relevance to selecting the model checkpoint? And if I have too many checkpoints, is manual selection still feasible?

**Comment by nianlonggu** [comment]
Yes, it would help. Just specify the log_file_name to where the training log file is located.
It works only if you have not modified the format of the logging output. Otherwise, you can write a new similar script to process the log file.

It only helps to find the best checkpoint number, not to automatically remove other checkpoints, etc.

**Comment by nhatdtn** [comment]
"Can you help me answer these two questions?

1. During training, the terminal displays multiple lines calculating validation scores, but I'm unsure which line corresponds to the validation score of each checkpoint.
2. The log file doesn't show the validation scores for each checkpoint consistently. Sometimes it only displays the scores for a few random checkpoints. Is this an error, and if so, why does it happen?"

**Comment by nianlonggu** [comment]
> "Can you help me answer these two questions?
> 
> 1. During training, the terminal displays multiple lines calculating validation scores, but I'm unsure which line corresponds to the validation score of each checkpoint.
> 2. The log file doesn't show the validation scores for each checkpoint consistently. Sometimes it only displays the scores for a few random checkpoints. Is this an error, and if so, why does it happen?"

Did you get a log file like below?

.....
[current_batch: 00900] loss: 1.691, learning rate: 0.000100
[current_batch: 01000] loss: 1.681, learning rate: 0.000100
Starting validation ...
val: 0.4215, 0.1711, 0.3839
[current_batch: 01100] loss: 1.679, learning rate: 0.000100
.....
[current_batch: 02000] loss: 1.634, learning rate: 0.000100
Starting validation ...
val: 0.4464, 0.1938, 0.4077
.....

The line starting with "val:" represents the validation loss, and the "current_batch" number right before it represents the current training steps of the model being validated.
There is a script "[get_optimal_batch.py](https://github.com/nianlonggu/MemSum/blob/main/src/MemSum_Full/get_optimal_batch.py)" that can help extract the validation performance and detect the best checkpoint on validation set. You can also write your own regular expression parser, or modify the train.py to save the validation results in a separate file: https://github.com/nianlonggu/MemSum/blob/50a507b0d24ad49fc0e8bca78903858a7c9ca2b8/src/MemSum_Full/train.py#L398

About the second question, I don't know what you mean exactly. The validation loss will be reported every "validate_every" steps, or at the end of an epoch, so you will see some reported checkpoints's batch_number is not a multiple of "validate_every". You can modify this following line to change the validation behavior:
https://github.com/nianlonggu/MemSum/blob/50a507b0d24ad49fc0e8bca78903858a7c9ca2b8/src/MemSum_Full/train.py#L383C14-L383C28

**Comment by nhatdtn** [comment]
My log file looks like this, it only calculates the validation score after current_batch_1000 and current_batch_1100. What about the remaining current_batches?
![image_2023_07_04T03_30_14_339Z](https://github.com/nianlonggu/MemSum/assets/135865937/ea6aa1c1-9ec0-4087-8da4-6fc88004e114)

**Comment by nianlonggu** [comment]
> My log file looks like this, it only calculates the validation score after current_batch_1000 and current_batch_1100. What about the remaining current_batches? ![image_2023_07_04T03_30_14_339Z](https://user-images.githubusercontent.com/135865937/250757132-ea6aa1c1-9ec0-4087-8da4-6fc88004e114.png)

In your case, one epoch contains 1100 batches, after 1100 batches the training goes to the next epoch, and you will see another validation loss at current batch = 2000.  Please see line: https://github.com/nianlonggu/MemSum/blob/50a507b0d24ad49fc0e8bca78903858a7c9ca2b8/src/MemSum_Full/train.py#L383C14-L383C28

**Comment by nhatdtn** [comment]
> 

So, the line "val: 0.4440, 0.1737, 0.3123" belongs to current_batch_1100, and then the line "val: 0.4553, 0.1474, 0.2779" would be from current_batch_1000, right? Because my training process stopped at current_batch_1500, as shown in the image, which represents my entire log file.

**Comment by nianlonggu** [comment]
Yes. 
You can validate more frequently by setting validate_every to a smaller value or increasing num_of_epochs

**Comment by nhatdtn** [comment]
Oh, I understand now. So, the calculation at batch_1000 represents the step determined by validate_every, while the calculation at batch_1100 marks the end of an epoch, which is why it continues to calculate the validation score. 
Thank you very much for your patient assistance!!
